cf/weights.py

```diff
@@ -0,0 +1,1903 @@
+import cfdm
+
+from .mixin_container import Container
+
+
+class Weights(Container, cfdm.Container):
+    """Worker functions for `cf.Field.weights`.
+
+    .. versionadded:: 3.16.0
+
+    """
+
+    @classmethod
+    def area_XY(
+        cls,
+        f,
+        weights,
+        weights_axes,
+        auto=False,
+        measure=False,
+        radius=None,
+        methods=False,
+    ):
+        """Calculate area weights from X and Y dimension coordinate
+        constructs.
+
+        .. versionadded:: 3.16.0
+
+        :Parameters:
+
+            {{weights weights: `dict`}}
+
+            {{weights weights_axes: `set`}}
+
+            {{weights auto: `bool`, optional}}
+
+            {{weights measure: `bool`, optional}}
+
+            {{radius: optional}}
+
+            {{weights methods: `bool`, optional}}
+
+        :Returns:
+
+            `bool`
+                `True` if weights were created, otherwise `False`.
+
+        """
+        from .units import Units
+
+        xkey, xcoord = f.dimension_coordinate(
+            "X", item=True, default=(None, None)
+        )
+        ykey, ycoord = f.dimension_coordinate(
+            "Y", item=True, default=(None, None)
+        )
+
+        if xcoord is None or ycoord is None:
+            if auto:
+                return False
+
+            raise ValueError(
+                "No unique X and Y dimension coordinate constructs for "
+                "calculating area weights"
+            )
+
+        radians = Units("radians")
+        metres = Units("metres")
+
+        if xcoord.Units.equivalent(radians) and ycoord.Units.equivalent(
+            radians
+        ):
+            pass
+        elif xcoord.Units.equivalent(metres) and ycoord.Units.equivalent(
+            metres
+        ):
+            radius = None
+        else:
+            if auto:
+                return False
+
+            raise ValueError(
+                "Insufficient coordinate constructs for calculating "
+                "area weights"
+            )
```

Comment:
```suggestion
        if xcoord.Units.equivalent(metres) and ycoord.Units.equivalent(
            metres
        ):
            radius = None
        elif not (xcoord.Units.equivalent(radians) and ycoord.Units.equivalent(
            radians
        )):
            if auto:
                return False

            raise ValueError(
                "Insufficient coordinate constructs for calculating "
                "area weights"
            )
```

to better (IMO) outline the logical cases where some action is being taken.

Reply:
I'm afraid I disagree in this case! I do agree that `if ...: pass` is a bit reductive, but I think it makes it clearer that the code in the `else` block happens when neither of two positive things have happened. Also (and this was the original reason for doing it like this), I would like to keep the radians test first, as it is will be more commonly passed, thereby avoiding the need for also doing the metres test in most cases.

Reply:
(although I would normally agree in principle :))